standard/attributes.md

```diff
@@ -858,6 +860,277 @@ For invocations that occur within field or event initializers, the member name u
+> ```csharp
+> public class X
+> {
+>     [DoesNotReturn]
+>     private void FailFast() =>
+>         throw new InvalidOperationException();
+>
+>     public void SetState(object? containedField)
+>     {
+>         if ((!isInitialized) || (containedField == null))
+>         {
+>             FailFast();
+>         }
+>         // null check not needed.
+>         _field = containedField;
+>     }
+> 
+>     private bool isInitialized = false;
+>     private object _field;
+> }
+> ```
+>
+> The presence of the attribute helps the compiler in a number of ways. First, the compiler can issue a warning if there's a path where the method can exit without throwing an exception. Second, the compiler can suppress nullable warnings in any code after a call to that method, until an appropriate catch clause is found. Third, the unreachable code won't affect any null states.
+>
+> The compiler can't change reachability (§13.2) or definite assignment (§9.4) analysis based on the presence of this attribute. A compiler can use this attribute only to impact nullability warnings. *end example*
+
+#### §The-DoesNotReturnIf-Attribute The DoesNotReturnIf attribute
+
+Specifies that a given method never returns if the associated `bool` parameter has the specified value.
+
+> *Example*: Consider the following:
+>
+> <!-- Example: {template:"standalone-lib", name:"DoesNotReturnIfAttribute", expectedWarnings:["CS0414"]}  -->
+> ```csharp
+> #nullable enable
+> public class X
+> {
+>     private void ThrowIfNull([DoesNotReturnIf(true)] bool isNull, string argumentName)
+>     {
+>         if (!isNull)
+>         {
+>             throw new ArgumentException(argumentName, $"argument {argumentName} can't be null");
+>         }
+>     }
+>
+>     public void SetFieldState(object containedField)
+>     {
+>         ThrowIfNull(containedField == null, nameof(containedField));
+>         // unreachable code when "isInitialized" is false:
+>         _field = containedField;
+>     }
+> 
+>     private bool isInitialized = false;
+>     private object _field = default!;
+> }
+> ```
+>
+> *end example*
+
+#### §The-MaybeNull-Attribute The MaybeNull attribute
+
+Specifies that a non-nullable return value may be null.
+
+> *Example*: Consider the following generic method:
+>
+> <!-- Example: {template:"code-in-class-lib-without-using", name:"MaybeNull1Attribute", replaceEllipsis:true, customEllipsisReplacements: ["return default;"]} -->
+> ```csharp
+> #nullable enable
+> public T? Find<T>(IEnumerable<T> sequence, Func<T, bool> predicate) { ... }
+> ```
+>
+> If `T` is replaced by `string`, `T?` becomes a nullable annotation. If `T` is replaced by `int`, `T?` becomes an `int?`. When `Find` searches an `IEnumerable<string>`, the default return is `null`, but when `Find` searches an `IEnumerable<int>`, the default return is 0. As such, specifying the return type as `T?` isn’t appropriate. However, adding this attribute solves the problem:
```

Comment:
The example doesn't compile in 8, but in 9 where it does, `T?` does not become `int?`; it stays as `int`. It can be thought of as "defaultable" rather than "nullable" in this context, when unconstrained.